### PR TITLE
[opt](Nerieds) add infer props to expression

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

mark this expression which is from predicate infer

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 1]
```diff
@@ -130,7 +132,7 @@ private Expression doInfer(ComparisonInferInfo equalInfo, ComparisonInferInfo pr
                 .comparisonPredicate.withChildren(newLeft, newRight);
         Expression expr = SimplifyComparisonPredicate.INSTANCE
                 .rewrite(TypeCoercionUtils.processComparisonPredicate(newPredicate), null);
-        return DateFunctionRewrite.INSTANCE.rewrite(expr, null);
+        return DateFunctionRewrite.INSTANCE.rewrite(expr, null).withInferred(true);
     }
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:136]: why not put it in construct func with default value

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java (hunk 1 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:136]: yeah, it's a good idea to construct it, i have fixed it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 2]
```diff
@@ -88,6 +88,8 @@ public ComparisonInferInfo(InferType inferType,
     public Set<Expression> infer(Set<Expression> predicates) {
         Set<Expression> inferred = Sets.newHashSet();
         for (Expression predicate : predicates) {
+            // if we support more predicate infer, we should add .withInferred(this.isInferred())
+            // to mark the predicate is from infer when call withChildren method
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:92]: mark: remove it

**Review by keanji-x — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java [hunk 3]
```diff
@@ -216,11 +232,19 @@ public int getDepth() {
...
+
     @Override
     public Expression withChildren(List<Expression> children) {
         throw new RuntimeException();
     }
 
+    public Expression withInferred(boolean inferred) {
+        throw new RuntimeException();
+    }
+
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java:247]: Add some error message

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**